### DISCUSSION TITLE
How do I catch ValidationError

### DISCUSSION THREAD

**Discussion by sky-183** [fastapi/fastapi#2993] [Questions] [answered]
Hello!
In my project I need to validate a list of objects and show list of bad ones + their errors if any ValidationError happened or list of good ones. And upload to database all good ones in any case.

Right now I cannot figure out, how to do it, since I cannot use try-except inside a path function to do it in a good way. One bad way to do it would be to accept json as it is (without pydantic's request model), then parse and feed it to pydantic one by one inside a try-except block to accumulate those good and bad lists of objects and errors. But this will ruin documentation, so I was thinking, "There must be a better way!"

I know, that method "add_exception_handler" can alter ValidationError's behavior, but I cant figure out, how to use it to achieve my goal, only to alter output a bit.

**Comment by ghandic** [comment, marked as answer]
I have two suggestions (depending on if you want your model in the body or query)

## Body

Essentially we're creating a dummy class that skips all validation and saves it for your dependency injection to handle

```python
from typing import Any, List, Optional

from fastapi import FastAPI, Depends
from pydantic import BaseModel, ValidationError
from pydantic.main import ModelMetaclass

app = FastAPI()


class AllOptional(ModelMetaclass):
    def __new__(self, name, bases, namespaces, **kwargs):
        annotations = namespaces.get("__annotations__", {})
        for base in bases:
            annotations = {**annotations, **base.__annotations__}
        for field in annotations:
            if not field.startswith("__"):
                annotations[field] = Optional[annotations[field]]
        namespaces["__annotations__"] = annotations
        return super().__new__(self, name, bases, namespaces, **kwargs)


class Animal(BaseModel):
    name: str
    species: str


class _Animal(Animal, metaclass=AllOptional):
    ...


class Animals(BaseModel):
    good: List[Animal] = []
    bad: List[Any] = []


def try_best_animals_body(animals: List[_Animal]) -> Animals:
    all_animals = Animals()
    for animal in animals:
        try:
            all_animals.good.append(Animal(**animal.dict()))
        except ValidationError:
            all_animals.bad.append(animal)
    return all_animals


@app.post("/animal-body")
def upload_animals(animals: Animals = Depends(try_best_animals_body)):
    return animals
```

## Query

```python
from typing import Any, List
from itertools import zip_longest

from fastapi import FastAPI, Depends, Query
from pydantic import BaseModel, ValidationError

app = FastAPI()


class Animal(BaseModel):
    name: str
    species: str


class Animals(BaseModel):
    good: List[Animal] = []
    bad: List[Any] = []


def try_best_animals_query(
    name: List[str] = Query([], alias="animal[name]"), species: List[str] = Query([], alias="animal[species]")
) -> Animals:
    all_animals = Animals()
    for animal in zip_longest(name, species):
        try:
            all_animals.good.append(Animal(name=animal[0], species=animal[1]))
        except ValidationError:
            all_animals.bad.append(animal)
    return all_animals


@app.post("/animal-query")
def upload_animals(animals: Animals = Depends(try_best_animals_query)):
    return animals

```


You're still having to handle your validation logic yourself as a try-except (mainly because it is application-specific), but you'll get your documentation with it this way

If this helped then could you mark it as answered? If not then happy to discuss further 😄

**Comment by tiangolo** [comment]
I think it would probably be better to ask in a GitHub issue, following the template that will guide you to structure the question in a way that makes it easier to understand your intention and need, to be able to help you better. And also providing a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) that I can copy-paste to replicate it.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by caeser1996** [comment]
You can create a custom exception handler for the ValidationError and modify the behavior of the validation error to return the list of good and bad objects and their errors.

Here's an example of how you can implement it:
```
from fastapi import FastAPI, HTTPException
from pydantic import ValidationError
from typing import List

app = FastAPI()

@app.exception_handler(ValidationError)
async def validation_error_handler(request, exc):
    good_objects = []
    bad_objects = []
    errors = {}
    
    # Iterate over the list of objects
    for index, obj in enumerate(exc.raw_value):
        try:
            # Validate the object using pydantic
            validated_obj = YourModel(**obj)
            good_objects.append(validated_obj)
        except ValidationError as e:
            # If validation error, add the object to bad objects
            # and the error message to errors dictionary
            bad_objects.append(obj)
            errors[index] = e.messages
    
    # Upload all good objects to the database
    upload_to_database(good_objects)
    
    # Return the response with the list of good and bad objects
    # and their errors, if any
    return {"good_objects": good_objects, "bad_objects": bad_objects, "errors": errors}

@app.post("/objects")
async def create_objects(objects: List[YourModelIn)]:
    # Your code here
    return objects

```
In this example, the custom exception handler will be called whenever a ValidationError is raised. It will loop over the list of objects, validate each object using Pydantic, and separate the good objects from the bad ones. The good objects will be uploaded to the database, and the response will include the list of good and bad objects and their errors, if any.

This approach allows you to validate the objects, separate the good and bad ones, and upload the good ones to the database all in one go, while still providing a clean and well-documented API using Pydantic's request models.
